EuiDualRange component

thompsongl opened the pull request:
### Summary

Resolves #496 

* Refactor `EuiRange` into smaller, shareable components (**no breaking changes unless you count class names**)
* New `EuiDualRange` that uses `EuiRange[X]` components
* Move Range docs/examples out of form controls section


### Checklist

- [x] Incorporate changes from #1461

- [x] This was checked in mobile
- [x] This was checked in IE11
- [x] This was checked in dark mode
- [x] Any props added have proper autodocs
- [x] Documentation examples were added
- [x] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately
- [x] This was checked for breaking changes and labeled appropriately
- [x] Jest tests were updated or added to match the most common scenarios
- [x] This was checked against keyboard-only and screenreader scenarios

~~- [ ] This required updates to Framer X components~~

thompsongl commented:
@nreese @thomasneirynck Let me know your thoughts on the API, specifically the returned `value` array for the 2 values. The actual `input` element will also have the 2 values reflected in its `value` attribute, but it'll appear as comma separated (we're "extending" the native behavior of the underlying form element here)

thomasneirynck commented:
@thompsongl love it!

Array with 2 numbers works for me. 

As for the input-element having the value, that's fine imho. That is also probably something we do not really have to worry about from a user perspective, correct? (ie. that the dual-range component is implemented using `<input>` seems only really relevant for testing).

thompsongl commented:
@thomasneirynck Great!
Yes, the use of `input` is _mostly_ an implementation detail, but I didn't want to alienate folks who rely heavily on native `form` events (e.g., submit). I think in most cases the React prop value will be consumed

cchaos commented:
@thompsongl Let me know when you've been able to
> Incorporate changes from #1461

Then I can do a more thorough review.

At first glance (not at the code) I just saw a couple of issues:

1. I think the dual range should always `showRange`
<img src="https://d.pr/free/i/FOmiMZ+" />

2. The width of the minimum input resizes when dragging the handles
<img src="https://d.pr/free/i/wqqgTb+" />

thompsongl commented:
@cchaos will do.

1. Good call on `showRange`. I'll default to `true` there.

2. This happens because the `max` value for the lower range value is the upper range value. There was a `maxWidth` calculation in there from `EuiRange` based on the number of digits in `max`. When the upper range value changes from 3 digits to 4 digits, the size changes. To avoid the resize, we'd probably be looking at erring on the side of larger inputs. Does that work?

cchaos commented:
2. Yes, that's correct. Whatever is the largest possible length of the value – meaning the most amount of characters.

nreese commented:
@thompsongl Looks really nice. Need some error handling for when `showInputs` is true that does not show the range, or range handles when max is less then min

<img width="495" alt="screen shot 2019-01-29 at 8 07 23 am" src="https://user-images.githubusercontent.com/373691/51917574-0c9d8c00-239d-11e9-83be-e170d9073ba8.png">

Since this PR refactors EuiRange so much. Should I just close https://github.com/elastic/eui/pull/1461 and let you migrate those ideas in this PR?

thompsongl commented:
@nreese Yeah, I have a task in here to incorporate changes from that PR. I'm fine pulling those changes into this changeset if #1461 won't merge soon. I'm also fine dealing with the merge conflicts if we want the commit history from your work.

thompsongl commented:
@nreese @chandlerprall @cchaos I've incorporated the concepts from #1460 and addressed all other feedback.

Re: Validation:
* The range/highlight track is now hidden for invalid values
* Tooltip is conditionally hidden for empty string value
* `isInvalid` is passed to parent components via `onChange` (implementation differs between range and dualRange because of reliance on `event`)

cchaos commented:
@thompsongl 

The other thing that #1461 did was to allow the inputs to be completely empty. Right now when you delete the contents it always reverts to `0` and doesn't allow the user to start typing without a beginning `0` unless they move the cursor.

Your PR:

<img src="https://d.pr/free/i/s4HfJ0+" width="50%" />

His PR:

<img src="https://d.pr/free/i/wBNGyU+" width="50%" />

---

I will make a note to myself that we need to flesh out the docs and break apart the examples for this component. But that can come later so that we can unblock the Kibana implementation.

cchaos commented on src/services/number/number.test.tsx:
```diff
@@ -0,0 +1,19 @@
+    expect(isWithinRange(10, 100, 0)).toBe(false);
+    expect(isWithinRange(0, 100, '')).toBe(false);
+    expect(isWithinRange(0, 100, null)).toBe(false);
+    expect(isWithinRange(0, 100)).toBe(false);
```
You might want to add some negative numbers in here too. Like `-10 to 10`?

thompsongl commented:
@cchaos Matched the behavior from the other PR regarding `0` vs. `''` in inputs.

nreese commented:
I found one issue. In EuiDualRange, set the min in the text to box to a large number, like 5000. Then try to grab the max slider.

thompsongl commented:
@nreese This is an interesting case. In terms of interaction, I'm not exactly sure how we want to handle a user _intentionally_ entering themselves into an invalid state. The only thing that makes sense to me at this time is to automatically reenter them into a valid state once they interact with the slider thumbs (e.g., change 5000 to the max value allowed by the slider).

I hesitate to create different state values for elements within the DualRange (e.g., a valid value for thumb position and an invalid value for the corresponding `input`) Reconciliation gets tricky, because the source of truth is obscured. Likewise, I hesitate to swap lower and upper values based on intentionally invalid user input; it's possible they aren't done with their selections.

@cchaos or @chandlerprall do you have input here?

cchaos commented:
Hmm, good question...

> The only thing that makes sense to me at this time is to automatically reenter them into a valid state once they interact with the slider thumbs 

I think that makes the most sense to me

chandlerprall commented:
Could also fire the onChange handler in componentDidUpdate if min > max

thompsongl commented:
Added logic for sane value resets if the user has gotten themselves into an invalid state and attempts to interact with the slider thumbs

nreese commented:
Just tried out the latest and it looks great. Thanks for making this

chandlerprall commented:
### right arrow key locks up
Sometimes the right arrow key doesn't work on the max slider. I've been able to hit this state a few different ways, the only way I can reproduce 100% is to go to the last dual range example, modify the max value via mouse, tab backwards to the previous range's max slider, and press arrow key. Nothing happens, even after repeated key strokes. Left arrow key works as expected, and then the right arrow works.

![right key prevented](https://d.pr/i/qJq6kt.gif)

### keyboard stepping
Steps don't work properly when using arrow keys. The range's step value is 50, from the example range's default state of 120 the right arrow key jumps to 170 and then 220. Dragging with the mouse aligns the step increments with the min value (`value % step === 0`).

![range stepping](https://d.pr/i/RZiMin.gif)

The HTML5 range control always steps to the proper increments (the native range can't be set to an invalid value, which probably helps it here).

![html5 range stepping](https://d.pr/i/gOVjs8.gif)

### page refresh is prevented if focused on a thumb
The `preventDefault` call prevents page refreshes and any other actions (e.g. form submit via enter). It should probably only happen for the arrow key intercepts.

### thumb swap doesn't snap to expected value
When dragging one thumb across the other, the swapped values aren't what I expect. E.g. if there's a max of 1050 and I drag the min thumb over 1050, I expect the new min to be 1050, instead the value is always lower (depending on the amount of pixels jumped by the last mouse movement).

![thumb swap](https://d.pr/i/Jbd1C9.gif)

thompsongl commented:
@chandlerprall I've fixed the first 3 of your comments. Good catch on that front!

I'm working through options for the last one.

thompsongl commented:
Updates for the last comment from @chandlerprall

chandlerprall commented:
@snide @cchaos thoughts on the modified classes? e.g. https://github.com/elastic/eui/pull/1485/files#diff-7ff6102775c5757cab5eb3e9443c3bff

cchaos commented:
Hmm, there are some class changes that look concerning especially in the jest files, though I thought they looked right when I reviewed (like compressed). I'll have to take another look.

chandlerprall commented on src/components/form/range/dual_range.js:
```diff
@@ -0,0 +1,433 @@
+  _handleKeyDown = (value, e) => {
+    let newVal = Number(value);
+    let changeRemainder = 0;
+    const change = this.props.step || 1;
```
rename to `step` for readability, and `changeRemainder` to `stepRemainder` - the whole operation is changing the value and it's not immediately clear what _change_ represents.

thompsongl replied:
Done

thompsongl commented on src/components/form/range/index.d.ts:
```diff
@@ -32,4 +34,19 @@ declare module '@elastic/eui' {
+    }
+
+    export const EuiDualRange: SFC<
+      CommonProps & InputHTMLAttributes<HTMLInputElement> & EuiRangeProps & EuiDualRangeProps
```
Fixed

thompsongl commented on src/components/form/range/index.d.ts:
```diff
@@ -32,4 +34,19 @@ declare module '@elastic/eui' {
+
+    export interface EuiDualRangeProps {
+      // Override acceptable value type
+      value: Array<number | string>
```
Fixed

thompsongl commented on src/services/number/number.ts:
```diff
@@ -0,0 +1,15 @@
+import { isNil } from '../predicate';
+
+export const isWithinRange = (min?: any, max?: any, value?: any) => {
```
All updated

chandlerprall commented on src/services/number/number.ts:
```diff
@@ -1,15 +1,7 @@
-export const isWithinRange = (min?: any, max?: any, value?: any) => {
-  if (isNil(value) || isNaN(value) || value === '') {
+export const isWithinRange = (min: number | string, max: number | string, value: number | string) => {
+  if (value === '') {
```
should check `min` and `max` for 0-length as well

thompsongl commented:
Re: class name changes:
I should've put a comment in there from the start, but it was a Zoom conversation with @snide that had me believe that class name changes (typically) do no constitute breaking changes. After the refactor, it was clear that there were a lot more BEM Block-level components than the Sass showed originally. React component == Sass Block was my thinking.

Could be that this isn't your concern, but more context for you all, nonetheless.

thompsongl commented on src/components/form/range/range_slider.js:
```diff
@@ -0,0 +1,59 @@
+    PropTypes.string,
+    PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.number, PropTypes.string]))
+  ]),
+  hasFocus: PropTypes.bool
```
`EuiSlider` relies on `:focus`, but `EuiDualRange` can't in tab/keyboard situations because it's not a _real_ form element. We have to simulate focus behavior for the track with local state, triggered when the thumbs have focus.

cchaos replied:
Ahh ok got it thx

thompsongl commented on src/components/form/range/range_track.js:
```diff
@@ -0,0 +1,186 @@
+  ),
+};
+
+const EuiRangeTicks = ({ disabled, onChange, ticks, tickObject, value, max }) => {
```
I debated this a couple different times. Ultimately, I left the single-dependency components in their parent files, but the reason doesn't hold up.
I'll move all the ones you commented on.
